File: cpp/tests/test_gather_statistics.cpp
/**
 * SPDX-FileCopyrightText: Copyright (c) 2025-2026, NVIDIA CORPORATION & AFFILIATES. All rights reserved.
 * SPDX-License-Identifier: Apache-2.0
 */

#include <memory>
#include <string>
#include <vector>

#include <gtest/gtest.h>

#include <rapidsmpf/coll/gather_statistics.hpp>
#include <rapidsmpf/statistics.hpp>
#include <rapidsmpf/utils/misc.hpp>

#include "environment.hpp"

using namespace rapidsmpf;

extern Environment* GlobalEnvironment;

TEST(GatherStatisticsTest, Basic) {
    auto const& comm = GlobalEnvironment->comm_;
    auto const rank = comm->rank();
    auto const nranks = comm->nranks();

    auto stats = std::make_shared<Statistics>();
    stats->add_stat("x", safe_cast<double>(rank));

    auto others = coll::gather_statistics(comm, 0, stats);

    if (rank == 0) {
        // Root receives nranks - 1 Statistics from other ranks.
        ASSERT_EQ(others.size(), safe_cast<std::size_t>(nranks - 1));
        for (auto const& s : others) {
            EXPECT_TRUE(s->enabled());
            EXPECT_EQ(s->list_stat_names().size(), 1);
            // Each non-root rank sent stat "x" with its rank value.
            EXPECT_EQ(s->get_stat("x").count(), 1);
        }
        // Merge and verify the total.
        auto global = stats->merge(others);
        // Sum of ranks: 0 + 1 + 2 + ... + (nranks-1) = nranks*(nranks-1)/2
        double expected_sum = safe_cast<double>(nranks) * (nranks - 1) / 2.0;
        EXPECT_EQ(global->get_stat("x").value(), expected_sum);
        EXPECT_EQ(global->get_stat("x").count(), safe_cast<std::size_t>(nranks));
    } else {
        EXPECT_TRUE(others.empty());
    }

    GlobalEnvironment->barrier();
}

TEST(GatherStatisticsTest, SingleRank) {
    auto const& comm = GlobalEnvironment->comm_;
    if (comm->nranks() != 1) {
        GTEST_SKIP() << "Test only meaningful with 1 rank";
    }

    auto stats = std::make_shared<Statistics>();
    stats->add_stat("x", 42.0);

    auto others = coll::gather_statistics(comm, 1, stats);
    EXPECT_TRUE(others.empty());
}

TEST(GatherStatisticsTest, DisjointStatNames) {
    auto const& comm = GlobalEnvironment->comm_;
    auto const rank = comm->rank();
    auto const nranks = comm->nranks();

    auto stats = std::make_shared<Statistics>();
    stats->add_stat("rank-" + std::to_string(rank), 1.0);

    auto others = coll::gather_statistics(comm, 2, stats);

    if (rank == 0) {
        ASSERT_EQ(others.size(), safe_cast<std::size_t>(nranks - 1));
        auto global = stats->merge(others);
        // Should have one stat per rank.
        EXPECT_EQ(global->list_stat_names().size(), safe_cast<std::size_t>(nranks));
        for (Rank r = 0; r < nranks; ++r) {
            EXPECT_EQ(global->get_stat("rank-" + std::to_string(r)).value(), 1.0);
        }
    } else {
        EXPECT_TRUE(others.empty());
    }

    GlobalEnvironment->barrier();
}

File: python/rapidsmpf/rapidsmpf/tests/test_gather_statistics.py
# SPDX-FileCopyrightText: Copyright (c) 2026, NVIDIA CORPORATION & AFFILIATES.
# SPDX-License-Identifier: Apache-2.0
"""Tests for gather_statistics."""

from __future__ import annotations

from typing import TYPE_CHECKING

from rapidsmpf.coll import gather_statistics
from rapidsmpf.statistics import Statistics

if TYPE_CHECKING:
    from rapidsmpf.communicator.communicator import Communicator


def test_gather_basic(comm: Communicator) -> None:
    stats = Statistics(enable=True)
    stats.add_stat("x", float(comm.rank))

    others = gather_statistics(comm, 0, stats)

    if comm.rank == 0:
        assert len(others) == comm.nranks - 1
        for s in others:
            assert s.enabled
            assert len(s.list_stat_names()) == 1
    else:
        assert len(others) == 0


def test_gather_single_rank(comm: Communicator) -> None:
    if comm.nranks != 1:
        return

    stats = Statistics(enable=True)
    stats.add_stat("x", 42.0)

    others = gather_statistics(comm, 1, stats)
    assert len(others) == 0


def test_gather_disjoint_names(comm: Communicator) -> None:
    stats = Statistics(enable=True)
    stats.add_stat(f"rank-{comm.rank}", 1.0)

    others = gather_statistics(comm, 2, stats)

    if comm.rank == 0:
        assert len(others) == comm.nranks - 1
        # Each remote rank contributed a uniquely named stat.
        all_names: set[str] = set(stats.list_stat_names())
        for s in others:
            all_names.update(s.list_stat_names())
        assert len(all_names) == comm.nranks
        for r in range(comm.nranks):
            assert f"rank-{r}" in all_names
    else:
        assert len(others) == 0
